### DISCUSSION TITLE
Is it possible to set dynamic value for middleware matcher so it only runs on develop environment?

### DISCUSSION THREAD

**Discussion by logantai24** [vercel/next.js#38392] [Help]
In Nextjs 12.2 middleware we can export a config to filter Middleware to run on specific paths. Is it possible to export different matcher base on the env, so the middleware only runs on development branches but not on production branches? The goal is to enable basic auth only on development branches, but not on production, without triggering unnecessary middleware functions call. Thank you!

Working => would not trigger middleware on home route
```
export const config = {
  matcher: '/somefakeroute',
}
```

Not Working => middleware still triggers even setting the CURRENT_ENV (or any other env) to production, it does not accept a dynamic value here
```
export const config = {
  matcher: process.env.CURRENT_ENV === "production" ? "/somefakeroute" : "/" 
}
```

**Comment by icyJoseph** [comment]
That's strange isn't it? 

The thing is that the middleware compiler doesn't work the way you'd expect.

For example, this works:

```js
export const config = {
  matcher: "/"
}
```

But this doesnt:

```js
export let config = {
  matcher: "/"
}
```

AFAIK, they look for specific syntax, on an Abstract Syntax Tree built from the middleware. So this is not necessarily evaluating by importing your `config` and then reading matcher from it, but rather, your middleware is parsed as an AST, and then a specific type of syntax is evaluated. I guess there are reasons related to how this code ends up running, for doing it this way (I guess to avoid dynamic code generation?)

A solution would be to check for the env var and the pathname in the function body:

```js
// <root>/middleware.js
import type { NextRequest } from "next/server";

export function middleware(request: NextRequest) {
  if (process.env.NODE_ENV === "development") {
    if (request.nextUrl.pathname.startsWith("/")) {
      // This logic is only applied to /about
    }
  } else {
    if (request.nextUrl.pathname.startsWith("/dashboard")) {
      // This logic is only applied to /dashboard
    }
  }
}

```

**Comment by logantai24** [comment]
Thanks @icyJoseph , but by doing so the middleware function would still be invoked to evaluate the env in production? VS not invoking when using the config way.

**Reply by icyJoseph** [reply]
Yes. It is a bit of a deadlock at the moment.

**Reply by icyJoseph** [reply]
https://github.com/vercel/next.js/issues/38461#issuecomment-1183265400

A bit more info on this matter.
